### PR TITLE
Font Library: Load/Unload the font face in browser when toggling the variants

### PR DESCRIPTION
**Pull request by arthur791004**
## What?

Load/Unload the font face in browser when toggling the variants.

## Why?

Referring to “[Changes in font-weight are not immediately reflected on the canvas, requiring a page reload for updates to be visible.](https://github.com/WordPress/gutenberg/issues/58764)”, we have to load/unload the specified font face in the browser when the user is managing the fonts.

## How?

Call the `loadFontFaceInBrowser` when the user is enabling a font-variant. When the user is disabling a font-variant, call the `unloadFontFaceInBrowser` instead.

Note that [Font faces that were added to the set using the CSS font-face rule remain connected to the corresponding CSS, and cannot be deleted](https://developer.mozilla.org/en-US/docs/Web/API/FontFaceSet/delete).

## Testing Instructions

* Go to the Site Editor
* Open the Font Library Modal by Styles > Typography > Fonts
* Install the “ABeeZee” fonts from the “Install Fonts” tab
* Create the following paragraph:
  * ABeeZee Regular. Set the Font Family to ABeeZee, and the Appearance to “Regular”
  * ABeeZee Regular Italic. Set the Font Family to ABeeZee, and the Appearance to “Regular Italic”
* Make sure the fonts are displayed correctly
* Open the Font Library Modal again to turn off the font variants of ABeeZee
* Save and refresh the page
* Open the Font Library Modal again
* Toggle the font variants of ABeeZee, and make sure the fonts are still displayed correctly

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/13596067/b9f62fab-39fa-4388-961c-83284317bb12

### CODE REVIEW

**Review by okmttdhr — APPROVED** [review]
✅ Confirmed;

- the font-weight gets reflected instantly after installing and applying it
- the font-weight gets reflected after deleting and refreshing the page (not instantly)

Am I correct in understanding that addressing the latter issue would necessitate a fundamental change to the font installation process in the Font Library?

**Comment by arthur791004** [comment]
> Am I correct in understanding that addressing the latter issue would necessitate a fundamental change to the font installation process in the Font Library?

I made a change, https://github.com/WordPress/gutenberg/pull/59066/commits/872c35fddcde2276c6ce7da901153b8b1739858c, to unload the deleted font faces as well so it should get reflected instantly 🙂

File: packages/edit-site/src/components/global-styles/font-library-modal/utils/index.js [hunk 1]
```diff
@@ -121,6 +121,38 @@ export async function loadFontFaceInBrowser( fontFace, source, addTo = 'all' ) {
 	}
 }
 
+/*
+ * Unloads the font face and remove it from the browser.
+ * It also removes it from the iframe document.
+ *
+ * Note that Font faces that were added to the set using the CSS @font-face rule
+ * remain connected to the corresponding CSS, and cannot be deleted.
```
Comment (matiasbenedetto) [inline packages/edit-site/src/components/global-styles/font-library-modal/utils/index.js:129]: I'm not sure what is the meaning of this comment

File: packages/edit-site/src/components/global-styles/font-library-modal/utils/index.js [hunk 2]
```diff
@@ -121,6 +121,38 @@ export async function loadFontFaceInBrowser( fontFace, source, addTo = 'all' ) {
...
+ */
+export function unloadFontFaceInBrowser( fontFace, addTo = 'all' ) {
```
Comment (matiasbenedetto) [inline packages/edit-site/src/components/global-styles/font-library-modal/utils/index.js]: if we are removing something using the parameter name `addTo` seems contradictory with the functionality. Should it be `removeFrom` instead?

**Review by matiasbenedetto — COMMENTED** [review]
Apart from a few inline minor comments, I'm wondering about these questions: Is this PR still relevant? Could you please define more clearly the problem it is fixing? I tested how it behaves in `trunk` vs this branch and couldn't find differences. It could be useful to post screencasts featuring the behavior of this fix and the behavior of trunk.

**Comment by okmttdhr** [comment]
> Apart from a few inline minor comments, I'm wondering about these questions: Is this PR still relevant? Could you please define more clearly the problem it is fixing? I tested how it behaves in `trunk` vs this branch and couldn't find differences. It could be useful to post screencasts featuring the behavior of this fix and the behavior of trunk.

As per my understanding, this PR addresses the second Observed Issue outlined in https://github.com/WordPress/gutenberg/issues/58764, where changes in font-weight aren't immediately visible on the canvas after installing or removing Google Fonts.

Upon testing with the latest version of trunk, I observed that while the issue seems to be resolved when a font is installed (probably in https://github.com/WordPress/gutenberg/issues/58765), the problem persists upon removing the font. 

trunk

https://github.com/WordPress/gutenberg/assets/5287479/c5199a05-c5cc-4a5f-a20a-f4ef588d0fc4

this PR

https://github.com/WordPress/gutenberg/assets/5287479/fb3df794-94b7-4a5a-bf87-176658683f03

**Comment by matiasbenedetto** [comment]
@okmttdhr. Thanks for the additional details. Is the video posted in https://github.com/WordPress/gutenberg/pull/59066#issuecomment-1965642487 from trunk or this branch? Could you record the same example using the one to compare?

**Comment by okmttdhr** [comment]
>Thanks for the additional details. Is the video posted in [#59066 (comment)](https://github.com/WordPress/gutenberg/pull/59066#issuecomment-1965642487) from trunk or this branch? Could you record the same example using the one to compare?

I updated the [comment](https://github.com/WordPress/gutenberg/pull/59066#issuecomment-1965642487) to include this PR's change :)

File: packages/edit-site/src/components/global-styles/font-library-modal/utils/index.js (hunk 1 above)
Comment (arthur791004) [reply packages/edit-site/src/components/global-styles/font-library-modal/utils/index.js:129]: It's from [MDN](https://developer.mozilla.org/en-US/docs/Web/API/FontFaceSet/delete). The fonts can be added to the FontFace set as follows, and the first method cannot be removed from the FontFace set.
* The CSS rule
* The `document.fonts.add`

**Review by matiasbenedetto — APPROVED** [review]
This seems ready to go now. 

I just made 3 small changes: 
- I replace a a function that was implemented again in this PR by a call to the existing function: https://github.com/WordPress/gutenberg/pull/59066/commits/a8b89381cdc62fb3725be74590a509f5e87598d3
- Removed an inaccurate comment: https://github.com/WordPress/gutenberg/pull/59066/commits/3e4f0643f8ab1337dd30ad7c3fdb1bca94b5d1fd
- Added a call to formatFontFaceName to make these changes work in Firefox too: https://github.com/WordPress/gutenberg/pull/59066/commits/ec32719bb68f052fce96bb2b0bedb313fdc62a3f

It looks good to me.
Thanks for working on this.

**Comment by getdave** [comment]
This PR is slated for inclusion in WP 6.5 during the RC phase as it fixes a important bug in the Fonts Library feature. 

Specifically this fixes the editor so that when a Font variant is activated/deactivated the editor will automatically reflect that change without requiring a reload.

This is critical to the feature as without it users could inadvertently have the wrong variants active/inactive and not notice the change until they reload the editor.

@matiasbenedetto or @arthur791004 May also be able to provide additional context.

**Comment by getdave** [comment]
I just cherry-picked this PR to the pick/wp-65-rc-2 branch to get it included in the next release: 59e7d64048

**Comment by arthur791004** [comment]
Thank you 👍

**Pull request merged**